parser/parser/cc/grammars/typedruby.ypp

```diff
@@ -1296,100 +1304,197 @@ lbrace_cmd_block_start:
+                | arg tEQQ error
+                    {
+                      $$ = driver.build.binaryOp(self, $1, $2, driver.build.error_node(self, @2.end, @3.end));
+                    }
                 | arg tNEQ arg
                     {
                       $$ = driver.build.binaryOp(self, $1, $2, $3);
                     }
+                | arg tNEQ error
+                    {
+                      $$ = driver.build.binaryOp(self, $1, $2, driver.build.error_node(self, @2.end, @3.end));
+                    }
                 | arg tMATCH arg
                     {
                       $$ = driver.build.match_op(self, $1, $2, $3);
                       DIAGCHECK();
                     }
+                | arg tMATCH error
+                    {
+                      $$ = driver.build.match_op(self, $1, $2, driver.build.error_node(self, @2.end, @3.end));
+                      DIAGCHECK();
+                    }
                 | arg tNMATCH arg
                     {
                       $$ = driver.build.binaryOp(self, $1, $2, $3);
                     }
+                | arg tNMATCH error
+                    {
+                      $$ = driver.build.binaryOp(self, $1, $2, driver.build.error_node(self, @2.end, @3.end));
+                    }
                 | tBANG arg
                     {
                       $$ = driver.build.not_op(self, $1, nullptr, $2, nullptr);
                     }
+                | tBANG error
+                    {
+                      $$ = driver.build.not_op(self, $1, nullptr, driver.build.error_node(self, @2.end, @2.end), nullptr);
+                    }
                 | tTILDE arg
                     {
                       $$ = driver.build.unary_op(self, $1, $2);
                     }
+                | tTILDE error
+                    {
+                      $$ = driver.build.unary_op(self, $1, driver.build.error_node(self, @2.end, @2.end));
+                    }
                 | arg tLSHFT arg
                     {
                       $$ = driver.build.binaryOp(self, $1, $2, $3);
                     }
+                | arg tLSHFT error
+                    {
+                      $$ = driver.build.binaryOp(self, $1, $2, driver.build.error_node(self, @2.end, @3.end));
+                    }
                 | arg tRSHFT arg
                     {
                       $$ = driver.build.binaryOp(self, $1, $2, $3);
                     }
+                | arg tRSHFT error
+                    {
+                      $$ = driver.build.binaryOp(self, $1, $2, driver.build.error_node(self, @2.end, @3.end));
+                    }
                 | arg tANDOP arg
                     {
                       $$ = driver.build.logicalAnd(self, $1, $2, $3);
                     }
+                | arg tANDOP error
+                    {
+                      $$ = driver.build.binaryOp(self, $1, $2, driver.build.error_node(self, @2.end, @3.end));
```

Comment:
This case and the case for `tOROP` are wrong. Note that successful parse results for `tANDOP` use `logicalAnd` and `logicalOr` from the builder.

When I printed the `desugar-tree` for your change, I noticed that it didn't desugar `&&` the way it should have.

Another thing to consider: add a third test case for each of the cases you added here to show what a successful parse result looks like in the exp file:

```ruby
    puts (t && A)
    puts (t && def) # error: unexpected token ")"
    puts (t &&) # error: unexpected token ")"
```

Doing this and using `desguar-tree` would have made catching this bug easier.

Reply:
This should be fixed now. I re-scanned the `desugar-tree` and didn't see anything problematic.